src/Pos/Subscription.hs

```diff
@@ -0,0 +1,177 @@
+                                          resolveDnsDomains)
+import           Pos.Slotting            (getLastKnownSlotDuration)
+
+data KnownRelay = Relay {
+      -- | When did we find out about this relay?
+      relayDiscovered :: UTCTime
+
+      -- | Was this relay reported last call to findRelays?
+    , relayActive :: Bool
+
+      -- | When was the last time it _was_ reported by findRelays?
+    , relayLastSeen :: UTCTime
+
+      -- | When did we last experience an error communicating with this relay?
+    , relayException :: Maybe (UTCTime, SomeException)
+    }
+
+type KnownRelays = Map NodeId KnownRelay
+
+activeRelays :: KnownRelays -> [NodeId]
+activeRelays = map fst . filter (relayActive . snd) . M.toList
+
+-- | We never expect relays to close the connection
+data RelayClosedConnection = RelayClosedConnection
+  deriving (Show)
+
+instance Exception RelayClosedConnection
+
+subscriptionWorkers
+    :: forall ssc ctx m. (WorkMode ssc ctx m, SscWorkersClass ssc)
+    => NetworkConfig -> DnsDomains -> ([WorkerSpec m], OutSpecs)
+subscriptionWorkers networkCfg dnsDomains = first (:[]) <$>
+    worker subscriptionWorkerSpec $ subscriptionWorker' networkCfg dnsDomains
+
+subscriptionWorkerSpec :: OutSpecs
+subscriptionWorkerSpec = toOutSpecs [ convH (Proxy @MsgSubscribe) (Proxy @Void) ]
+
+subscriptionWorker'
+    :: forall ssc ctx m. (WorkMode ssc ctx m)
+    => NetworkConfig -> DnsDomains -> SendActions m -> m ()
+subscriptionWorker' networkCfg dnsDomains sendActions =
+    loop M.empty
+  where
+    loop :: KnownRelays -> m ()
+    loop oldRelays = do
+      slotDur <- getLastKnownSlotDuration
+      now     <- liftIO $ getCurrentTime
+      peers   <- findRelays
+
+      let delayInterval :: Millisecond
+          delayInterval = max (slotDur `div` 4) (convertUnit (5 :: Second))
+
+          updatedRelays :: KnownRelays
+          updatedRelays = updateKnownRelays now peers oldRelays
+
+      -- Declare all active relays as a single list of alternative relays
+      updateKnownPeers $ \peersOld -> do
+        -- NOTE: The assumption is that a single part of the code is
+        -- responsible for maintaining the set of known peers. For behind NAT
+        -- nodes, this is the responsibility of the subscription worker; for
+        -- P2P/transitional nodes, this is the responsibility of the
+        -- Kademlia worker; and for statically known sets of peers (core nodes
+        -- and light wallets) this happens at queue initialization time.
+        -- Here we just check that nobody else interferred with the set of
+        -- peers, just as a sanity check.
+        let expectedOld = peersFromList [(NodeRelay, activeRelays oldRelays)]
+        if peersOld == expectedOld
+          then peersFromList [(NodeRelay, activeRelays updatedRelays)]
+          else error "Invariant violated in subscriptionWorker"
+
+      -- Subscribe only to a single relay (if we found one)
```

Comment:
Ultimately we need to make this configurable so that we can have the config specify how many relays to subscribe to (and similarly how to organise the relays into lists of known peers so we try to send to a certain number of relays).